src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs

```diff
@@ -5568,6 +5568,477 @@ await foreach (string value in pets)
             VerifyOperationTreeForTest<BlockSyntax>(source, expectedOperationTree);
         }
 
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/49267")]
+        public void AsyncForeach_StructEnumeratorWithDispose()
+        {
+            var compilation = CreateCompilation(@"
+#pragma warning disable CS1998 // async method lacks awaits
+using System.Threading.Tasks;
+class C
+{
+    static async Task Main()
+    /*<bind>*/{
+        await foreach (var i in new C())
+        {
+        }
+    }/*</bind>*/
+    public AsyncEnumerator GetAsyncEnumerator() => throw null;
+    public struct AsyncEnumerator
+    {
+        public int Current => throw null;
+        public async Task<bool> MoveNextAsync() => throw null;
+        public async ValueTask DisposeAsync() => throw null;
+    }
+}", targetFramework: TargetFramework.NetCoreApp30);
+
+            VerifyOperationTreeAndDiagnosticsForTest<BlockSyntax>(compilation, @"
+IBlockOperation (1 statements) (OperationKind.Block, Type: null) (Syntax: '{ ... }')
+  IForEachLoopOperation (LoopKind.ForEach, IsAsynchronous, Continue Label Id: 0, Exit Label Id: 1) (OperationKind.Loop, Type: null) (Syntax: 'await forea ... }')
+    Locals: Local_1: System.Int32 i
+    LoopControlVariable: 
+      IVariableDeclaratorOperation (Symbol: System.Int32 i) (OperationKind.VariableDeclarator, Type: null) (Syntax: 'var')
+        Initializer: 
+          null
+    Collection: 
+      IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: C, IsImplicit) (Syntax: 'new C()')
+        Conversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+        Operand: 
+          IObjectCreationOperation (Constructor: C..ctor()) (OperationKind.ObjectCreation, Type: C) (Syntax: 'new C()')
+            Arguments(0)
+            Initializer: 
+              null
+    Body: 
+      IBlockOperation (0 statements) (OperationKind.Block, Type: null) (Syntax: '{ ... }')
+    NextVariables(0)
+            ", DiagnosticDescription.None);
+
+            VerifyFlowGraphForTest<BlockSyntax>(compilation, @"");
+        }
+
+        [Fact]
+        public void Foreach_ExtensionGetEnumeratorWithParams()
```

Comment:
>Foreach_ExtensionGetEnumeratorWithParams [](start = 20, length = 40)

Do we have a similar test for VB?

Reply:
We didn't add extension foreach support to VB, did it already have support for the feature?

---
In reply to: [529091269](https://github.com/dotnet/roslyn/pull/49534#discussion_r529091269) [](ancestors = 529091269)

Reply:
> We didn't add extension foreach support to VB, did it already have support for the feature?

Extension methods were always considered for patterns like that in VB:
```
Module Ext
    <System.Runtime.CompilerServices.Extension>
    Function GetEnumerator(this As MyEnumerable, ParamArray array As Integer()) As System.Collections.Generic.IEnumerator(Of Integer)
    End Function
End Module

Class MyEnumerable
    Sub Test()
        For Each x In New MyEnumerable
        Next
    End Sub
End Class
```

---
In reply to: [533002410](https://github.com/dotnet/roslyn/pull/49534#discussion_r533002410) [](ancestors = 533002410,529091269)